### PR TITLE
Added out feature during pooling in train_all_params.

### PR DESCRIPTION
**Pull request by jstaker7**
Hi All,

I'm in the process of going though your paper and code, hoping to reproduce the reported results. I'm currently working through sections 3.1 (repro atom typing) and 3.2 (recover MM parameters) of the paper. If I'm not mistaken, the scripts `espaloma/app/train_multi_typing.py` and `espaloma/app/train_all_params.py` are intended to reproduce sections 3.1 and 3.2 of the paper, respectively, is that correct? In my tests I've only set n_epochs to 5000 as mentioned in the paper and kept everything else as defaults; is there other parameters I need to set to fully reproduce the work?

In running `espaloma/app/train_multi_typing.py` I'm getting accuracies below the 98% reported in the paper using the default data/settings provided. Any ideas on what I should do differently? Results:

Screen Shot 2020-11-05 at 10 16 40 AM

Similarly, running `espaloma/app/train_all_params.py` results in the following, where something looks suspicious with r2:

Screen Shot 2020-11-13 at 10 51 38 AM

Lastly, running `espaloma/app/train_all_params.py` out of the box resulted in a traceback and I had to make the included alteration. Please let me know if this correction seems sensible.

Thanks!

### CODE REVIEW

**Comment by maxentile** [comment]
Hi @jstaker7 ! Thank you for your interest in our work, and for contributing this fix. It's possible the scripts in `app` are not the ones to be looking at, thank you for bearing with us as we sort this out. Also, I think @yuanqing-wang had been developing parts of the atom type recovery task for section 3.1 in this repository https://github.com/yuanqing-wang/hgfp_scripts/tree/master/types -- it's possible that we missed including something important from that development into the espaloma repository. (We should update the readme to provide a more specific pointer to the code associated with each experiment.) @yuanqing-wang : are there other hyperparameter settings required for the atom type recovery task in section 3.1?

**Comment by yuanqing-wang** [comment]
Hello @jstaker7! Many thanks for your interest.

Apologies that we did a poor job sharing the scripts we actually used in the paper. The scripts in the `app` module was supposed to be deleted. I'm removing that to prevent further confusion.

In terms of reproducing this result itself, it requires some hyperparameter tuning and please see the script with the final choices that lead to the reported result attached here:

```
import torch

import espaloma as esp


def run(layer_name):
    # grab dataset
    # esol = esp.data.esol(first=20)
    ds = esp.data.zinc(first=1000).shuffle()

    # do some typing
    typing = esp.graphs.legacy_force_field.LegacyForceField('gaff-1.81')
    ds.apply(typing, in_place=True) # this modify the original data

    # split
    # NOTE:
    # I don't like torch-generic splitting function as it requires
    # specifically the volume of each partition and it is inconsistent
    # with the specification of __getitem__ method
    ds_tr, ds_te = ds.split([4, 1])

    ds_tr.save('ds_tr.th')
    ds_te.save('ds_te.th')

    # get a loader object that views this dataset in some way
    # using this specific flag the dataset turns into an iterator
    # that outputs loss function, per John's suggestion
    ds_tr = ds_tr.view('graph', batch_size=20)
    ds_te = ds_te.view('graph', batch_size=len(ds_te))

    # define a layer
    layer = esp.nn.layers.dgl_legacy.gn(layer_name)

    # define a representation
    representation = esp.nn.Sequential(
            layer,
            [32, 'leaky_relu', 32, 'leaky_relu', 32, 'leaky_relu']
    )

    # define a readout
    readout = esp.nn.readout.node_typing.NodeTyping(
            in_features=32,
            n_classes=100
    ) # not too many elements here I think?

    net = torch.nn.Sequential(
        representation,
        readout
    )


    exp = esp.TrainAndTest(
        ds_tr=ds_tr,
        ds_te=ds_te,
        net=net,
        metrics_te=[esp.metrics.TypingAccuracy()],
        n_epochs=500,
    )

    results = exp.run()

    print(esp.app.report.markdown(results))

    import pickle
    with open(layer_name + "_ref_g_test.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_test, f_handle)

    with open(layer_name + "_ref_g_training.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_training, f_handle)


if __name__ == '__main__':
    import sys
    run(sys.argv[1])
```

Again, sorry that this is not part of the repo yet. I was planning to arrange these scripts nicely for the actual paper submission. I was quite pleasantly surprised to learn that this project already attracted attention like yours at this early stage.

**Comment by maxentile** [comment]
Thanks @yuanqing-wang ! To run the script you attached, I think we need to pass a command-line argument corresponding to `layer_name`. I think the results in `hgfp_scripts` suggested "SAGEConv" did better than other layer types here. Could you comment on what `layer_name`s are permissible, and which should be chosen here?

**Comment by yuanqing-wang** [comment]
The permissible layer names include the ones implemented in `dgl`: https://docs.dgl.ai/en/0.4.x/api/python/nn.html

`SAGEConv` is indeed chosen to produce the published results.

**Comment by jchodera** [comment]
@yuanqing-wang @maxentile : Let's be sure to cut a release and provide an `environment.yml` for the coming bioRxiv v1 submission. We can then run the scripts one more time to produce the results that go into the bioRxiv.

**Comment by jstaker7** [comment]
@yuanqing-wang, @maxentile, thanks so much for the additional info and the script. @yuanqing-wang do you also have a copy of the script for section 3.2 that you are able to share? I'd also like to give that a go, if possible. Thanks!

**Comment by jstaker7** [comment]
Hi @yuanqing-wang Thank you for providing the script for section 3.1. I was able to run it successfully.

Screen Shot 2020-11-16 at 10 58 58 AM

Do you have a similar script for section 3.2 (reproducing the parameters)?

**Comment by jstaker7** [comment]
@yuanqing-wang 

Here's my attempt at adapting the typing script to section 3.2:

```
import sys
import torch
import pickle

import espaloma as esp


def run(layer_name="SAGEConv"):
    # grab dataset
    ds = esp.data.alkethoh().shuffle()

    # do some typing
    typing = esp.graphs.legacy_force_field.LegacyForceField('smirnoff99Frosst').parametrize
    ds.apply(typing, in_place=True) # this modify the original data

    # split
    # NOTE:
    # I don't like torch-generic splitting function as it requires
    # specifically the volume of each partition and it is inconsistent
    # with the specification of __getitem__ method
    ds_tr, ds_te = ds.split([4, 1])

    # get a loader object that views this dataset in some way
    # using this specific flag the dataset turns into an iterator
    # that outputs loss function, per John's suggestion
    ds_tr = ds_tr.view('graph', batch_size=20)
    ds_te = ds_te.view('graph', batch_size=len(ds_te))

    # define a layer
    layer = esp.nn.layers.dgl_legacy.gn(layer_name)

    # define a representation
    representation = esp.nn.Sequential(
            layer,
            [32, 'leaky_relu', 32, 'leaky_relu', 32, 'leaky_relu']
    )

    # define a readout
    readout = esp.nn.readout.janossy.JanossyPooling(
        in_features=32,
        config=[32, "tanh"],
        out_features={1: ["k"], 2: ["k"], 3: ["k"], 4: ["k"]},
    )

    net = torch.nn.Sequential(representation, readout)
    
    metrics_tr = [
        esp.metrics.GraphMetric(
            base_metric=torch.nn.L1Loss(),
            between=[param, param + "_ref"],
            level=term,
        )
        for param in ["k"]
        for term in ["n2", "n3"]
    ]

    metrics_te = [
        esp.metrics.GraphMetric(
            base_metric=base_metric,
            between=[param, param + "_ref"],
            level=term,
        )
        for param in ["k"]
        for term in ["n2", "n3"]
        for base_metric in [esp.metrics.rmse, esp.metrics.r2]
    ]

    exp = esp.TrainAndTest(
        ds_tr=ds_tr,
        ds_te=ds_te,
        net=net,
        metrics_tr=metrics_tr,
        metrics_te=metrics_te,
        n_epochs=5000,
    )

    results = exp.run()

    print(esp.app.report.markdown(results))

    with open(layer_name + "_ref_g_test.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_test, f_handle)

    with open(layer_name + "_ref_g_training.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_training, f_handle)


if __name__ == '__main__':
    
    run()
```

Though it doesn't seem to be working as expected:

n2_n3

Any ideas?

**Comment by jchodera** [comment]
@yuanqing-wang @maxentile : Can you folks take a look this week, and make sure the scripts necessary to reproduce the experiments you had carried out are in the repo?

**Comment by yuanqing-wang** [comment]
Hi @jstaker7, sorry for the delay. Somehow this thread isn't showing up on my GitHub page. I will make sure to check more often.

Apologies for all the confusions and inconveniences. Although `espaloma` has been developed in a open-source manner, we haven't reach the stage yet to provide the users with clear documentation and so forth. 

Your script looks good over all, except for two subtleties that have affect on the overall flow of the information.

First is that your `out_features` for `janossy_pooling` must contain not only `k` but `eq` for `n2`, which is bonds, and `n3`, which is angles. In terms of torsions, note that we assume the phases to be zero and the periodicity to be `range(1, 7)`, which is fine if you think about this as a Fourier expansion, but not okay if you want to match the MM torsion energy exactly. So at current stage we don't encourage to fit to MM directly with torsions. So your readout section could look like
```
    readout = esp.nn.readout.janossy.JanossyPooling(
        in_features=32,
        config=[32, "tanh"],
        out_features={
            1: ["k"], 
            2: ["k", "eq"], 
            3: ["k", "eq"],
    },
    )
```

Second, when it comes to energy calculation, you'd need to specify steps to calculate the reference and predicted energy. So the overall model would look like:

```
    net = torch.nn.Sequential(
            representation, 
            readout,
            esp.mm.geometry.GeometryInGraph(),
            esp.mm.energy.EnergyInGraph(terms=["n2", "n3"],),
            esp.mm.energy.EnergyInGraph(terms=["n2", "n3"], suffix='_ref'),
    )
```

If you make these changes I think the results would be similar to what we've reported.

**Comment by yuanqing-wang** [comment]
But for reproducibility I'm also attached the script which was responsible for generating the results reported.


```
# =============================================================================
# IMPORTS
# =============================================================================
import argparse
import os

import numpy as np
import torch

import espaloma as esp


def run(args):
    # define data
    data = getattr(esp.data, args.data)(first=args.first)

    # get force field
    forcefield = esp.graphs.legacy_force_field.LegacyForceField(
        args.forcefield
    )

    # param / typing
    operation = forcefield.parametrize

    # apply to dataset
    data = data.apply(operation, in_place=True)

    # apply simulation
    # make simulation
    from espaloma.data.md import MoleculeVacuumSimulation
    simulation = MoleculeVacuumSimulation(
        n_samples=100, n_steps_per_sample=10,
    )

    data = data.apply(simulation.run, in_place=True)

    # split
    partition = [int(x) for x in args.partition.split(":")]
    ds_tr, ds_te = data.split(partition)

    # batch
    ds_tr = ds_tr.view("graph", batch_size=args.batch_size)
    ds_te = ds_te.view("graph", batch_size=args.batch_size)

    # layer
    layer = esp.nn.layers.dgl_legacy.gn(args.layer)

    # representation
    representation = esp.nn.Sequential(layer, config=args.config)

    # get the last bit of units
    units = [int(x) for x in args.config if x.isdigit()][-1]

    print(args.janossy_config)

    janossy_config = []
    for x in args.janossy_config:
        if isinstance(x, int):
            janossy_config.append(int(x))

        elif x.isdigit():
            janossy_config.append(int(x))

        else:
            janossy_config.append(x)

    print(janossy_config)

    readout = esp.nn.readout.janossy.JanossyPooling(
        in_features=units, config=janossy_config,
    )

    net = torch.nn.Sequential(
            representation, 
            readout,
            esp.mm.geometry.GeometryInGraph(),
            esp.mm.energy.EnergyInGraph(terms=["n2", "n3"]),
            esp.mm.energy.EnergyInGraph(terms=["n2", "n3"], suffix='_ref'),
    )
   
    metrics_tr = [
        esp.metrics.GraphMetric(
            base_metric=torch.nn.MSELoss(),
            between=['u', "u_ref"],
            level="g",
        ),

        
        esp.metrics.GraphDerivativeMetric(
              base_metric=torch.nn.MSELoss(),
              between=["u", "u_ref"],
              level="g",
              weight=10.0,
        ),
    ]


    metrics_te = [
        esp.metrics.GraphMetric(
            base_metric=esp.metrics.r2,
            between=['u', 'u_ref'],
            level="g",
        ),
        esp.metrics.GraphMetric(
            base_metric=esp.metrics.rmse,
            between=['u', 'u_ref'],
            level="g",
        ),

    ]

    exp = esp.TrainAndTest(
        ds_tr=ds_tr,
        ds_te=ds_te,
        net=net,
        metrics_tr=metrics_tr,
        metrics_te=metrics_te,
        n_epochs=args.n_epochs,
        normalize=esp.data.normalize.NotNormalize,
        optimizer=lambda net: torch.optim.Adam(net.parameters(), 1e-3),
        device=torch.device('cuda:0'),
    )

    results = exp.run()

    print(esp.app.report.markdown(results))

    import os
    os.mkdir(args.out)

    with open(args.out + "/architecture.txt", "w") as f_handle:
        f_handle.write(str(exp))

    with open(args.out + "/result_table.md", "w") as f_handle:
        f_handle.write(esp.app.report.markdown(results))

    curves = esp.app.report.curve(results)

    for spec, curve in curves.items():
        np.save(args.out + "/" + "_".join(spec) + ".npy", curve)

    import pickle
    with open(args.out + "/ref_g_test.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_test, f_handle)

    with open(args.out + "/ref_g_training.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_training, f_handle)


    print(esp.app.report.markdown(results))

    import pickle
    with open(args.out + "/ref_g_test.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_test, f_handle)

    with open(args.out + "/ref_g_training.th", "wb") as f_handle:
        pickle.dump(exp.ref_g_training, f_handle)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("--data", default="alkethoh", type=str)
    parser.add_argument("--first", default=-1, type=int)
    parser.add_argument("--partition", default="4:1", type=str)
    parser.add_argument("--batch_size", default=8, type=int)
    parser.add_argument("--forcefield", default="smirnoff99Frosst", type=str)
    parser.add_argument("--layer", default="GraphConv", type=str)
    parser.add_argument("--n_classes", default=100, type=int)
    parser.add_argument(
        "--config", nargs="*", default=[32, "tanh", 32, "tanh", 32, "tanh"]
    )

    parser.add_argument(
        "--training_metrics", nargs="*", default=["TypingCrossEntropy"]
    )
    parser.add_argument(
        "--test_metrics", nargs="*", default=["TypingAccuracy"]
    )
    parser.add_argument(
        "--out", default="results", type=str
    )
    parser.add_argument("--janossy_config", nargs="*", default=[32, "leaky_relu"])

    parser.add_argument("--n_epochs", default=10, type=int)

    args = parser.parse_args()

    run(args)
```

**Comment by yuanqing-wang** [comment]
The `config` we used was `[128, "sigmoid", 128, "sigmoid", 128, "sigmoid"]`

**Comment by yuanqing-wang** [comment]
Again all of these will be part of the repo once things we clear things up. Sorry again for the confusion.

**Comment by jstaker7** [comment]
@yuanqing-wang thanks so much for providing the script! 

In attempting to run it I get a "Loss is Nan" error. As a test, I also ran the script with only first=40 and it ran successfully without error. Two possibilities come to mind: either one or more datapoints are not well defined resulting in nans in the input, or possibly gradients are exploding during training causing weights to go to nan. Any thoughts on what might be happening, or if you've seen this behavior during training? If nothing immediate comes to mind perhaps I'll wait until you've had a chance to update the repo as you've mentioned and see if the update fixes the issues.

**Comment by yuanqing-wang** [comment]
Hi @jstaker7, what data you ran it with and what does your hyperparameter look like?

**Comment by jstaker7** [comment]
@yuanqing-wang I used the script and data as provided. I isolated the issue and it appears to be the simulation producing bad xyz values during featurization. It was not consistently reproducible and happened infrequently at random. However, I was able to run the script eventually without any bad values and saved the dataset to disk so I could run experiments. Though, once the script ran to completion this was the output (at least r2 looks suspicious):

Screen Shot 2020-12-04 at 9 22 47 AM

**Comment by yuanqing-wang** [comment]
Ah OK so you used the default args. The default is meant to provide an example to run and finish shortly. If you use the aforementioned config and train for 5000 epochs, you should be able to get similar results.

**Comment by jstaker7** [comment]
To clarify, I did use the mentioned config, though I did not specify epochs but used the default, so I can try running with more epochs.

**Comment by jstaker7** [comment]
@yuanqing-wang Thanks! The script seems to be working now:

Screen Shot 2020-12-09 at 7 56 38 AM

Looks like only the potential energy is outputted here, and I'd imagine the plan is to also eventually include in the repo the necessary additions to also output the parameters? In any case, feel free to close out this issue and I'll keep an eye out for any additional updates and/or new scripts. Thanks!

**Comment by yuanqing-wang** [comment]
Thanks! I'll let you know when this package is in a nicer stage and when we publish our next version of preprint.
